### PR TITLE
Fix flatten threading issue by directly using reduce(...) NIO method

### PR DESCRIPTION
**Pull request by PopFlamingo**
This also improves the test by checking that order the order of the results will match the order of the futures in the input array as it has been specified in the docs.

### CODE REVIEW

File: Sources/Async/Future+Flatten.swift [hunk 1]
```diff
@@ -40,41 +40,33 @@ extension Collection where Element: FutureType {
...
+                    promise.succeed(result: value)
+                case .error(let error):
+                    promise.fail(error: error)
                 }
             }
         }
-
-        // return future result
-        return promise.futureResult
+        let futures = promises.map { $0.futureResult }
```
Comment (Lukasa) [inline Sources/Async/Future+Flatten.swift:76]: From line 50 to here seems to be a complete no-op: you're building a new collection of `Future` objects that do nothing but carry the result of the input set. I think you can just replace this by writing `let futures = Array(self)`

File: Sources/Async/Future+Flatten.swift [hunk 2]
```diff
@@ -40,41 +40,33 @@ extension Collection where Element: FutureType {
...
+        return Future<[Element.Expectation]>.reduce(into: [],
```
Comment (siemensikkema) [inline Sources/Async/Future+Flatten.swift]: Just a Vapor style nitpick: When splitting functions with multiple argument over several lines, please start the first argument label on a new line (and indented 1 level) and have subsequent arguments aligned along the first.

File: Sources/Async/Future+Flatten.swift [hunk 3]
```diff
@@ -40,41 +40,33 @@ extension Collection where Element: FutureType {
...
+                                             futures,
+                                             eventLoop: eventLoop) { partialResult, nextElement in
+            return partialResult.append(nextElement)
+        }
+
```
Comment (siemensikkema) [inline Sources/Async/Future+Flatten.swift]: unnecessary newline

**Review by siemensikkema — COMMENTED** [review]
Just some style comments. Would be great if we could automate that ...

File: Sources/Async/Future+Flatten.swift [hunk 4]
```diff
@@ -40,41 +40,33 @@ extension Collection where Element: FutureType {
...
-                    // insert the element into its array index
-                    results[i] = el
-                    // zero remaining, succeed the promise
-                    if remaining == 0 {
-                        promise.succeed(result: results.compactMap { $0 })
-                    }
+        
+        var promises = [EventLoopPromise<Element.Expectation>]()
+        for future in self {
+            let promise = eventLoop.newPromise(of: Element.Expectation.self)
```
Comment (weissi) [inline Sources/Async/Future+Flatten.swift:65]: just a quick comment: `newPromise(of:)` was introduced with NIO 1.12 so Vapor's `Package.swift` will need an update to have `from: "1.12.0"` for NIO

File: Sources/Async/Future+Flatten.swift (hunk 3 above)
Comment (PopFlamingo) [reply Sources/Async/Future+Flatten.swift]: Thank you! I now it is fixed!

File: Sources/Async/Future+Flatten.swift (hunk 4 above)
Comment (PopFlamingo) [reply Sources/Async/Future+Flatten.swift:65]: I updated it! Thank you!

File: Sources/Async/Future+Flatten.swift (hunk 1 above)
Comment (tanner0101) [reply Sources/Async/Future+Flatten.swift:76]: Agreed. I think `flatten` here is really just a convenience for calling `reduce(into: [], self, ...)`

**Review by tanner0101 — CHANGES_REQUESTED** [review]

File: Sources/Async/Future+Flatten.swift [hunk 5]
```diff
@@ -40,41 +40,33 @@ extension Collection where Element: FutureType {
...
             }
         }
-
-        // return future result
-        return promise.futureResult
+        let futures = promises.map { $0.futureResult }
+        return Future<[Element.Expectation]>.reduce(
+            into: [],
+            futures,
+            eventLoop: eventLoop) { partialResult, nextElement in
```
Comment (tanner0101) [inline Sources/Async/Future+Flatten.swift]: closing `)` should be on a newline here.

File: Sources/Async/Future+Flatten.swift (hunk 1 above)
Comment (tanner0101) [reply Sources/Async/Future+Flatten.swift:76]: Edit: This extension is on `FutureType`, not `Future`, so we need to do it this way.

**Review by tanner0101 — APPROVED** [review]
Thanks!

File: Sources/Async/Future+Flatten.swift (hunk 5 above)
Comment (PopFlamingo) [reply Sources/Async/Future+Flatten.swift]: Fixed it! :)

**Comment by PopFlamingo** [comment]
@tanner0101 Added a stress test, it failed every time I ran it with the old buggy implementation.

**Comment by PopFlamingo** [comment]
@tanner0101 Maybe don't merge immediately, I noticed the memory usage was a little weird and would like to make sure everything is ok.

**Comment by PopFlamingo** [comment]
@tanner0101 After looking at the potential issue in detail, I couldn't find any actual cause of retain cycle, and with the help of @siemensikkema I discovered that the growing memory usage issue is gone when in release. No instrument detects a leak. I'm still concerned about this issue but maybe that's something expected when in debug?

**Comment by siemensikkema** [comment]
@adtrevor @tanner0101 the debug-mode memory leak is fixed here: https://github.com/apple/swift-nio/pull/752

**Comment by PopFlamingo** [comment]
@Lukasa Thanks a lot for fixing it and thank you for your support @siemensikkema ! :)

File: Tests/AsyncTests/AsyncTests.swift [hunk 6]
```diff
@@ -16,11 +16,54 @@ final class AsyncTests: XCTestCase {
...
         b.succeed(result: "b")
-        try XCTAssertEqual(flat.wait(), ["a", "b"])
+        a.succeed(result: "a")
+        c.succeed(result: "c")
+        try XCTAssertEqual(flat.wait(), ["a", "b", "c"])
+    }
+    
+    /// Stress test flatten
+    /// If this test fails, this might indicate a threading issue
+    func testFlattenStress() throws {
```
Comment (tanner0101) [inline Tests/AsyncTests/AsyncTests.swift:30]: I think this needs to be added to the `allTests` at the bottom of the file to run on Linux.

**Review by tanner0101 — CHANGES_REQUESTED** [review]

File: Tests/AsyncTests/AsyncTests.swift (hunk 6 above)
Comment (PopFlamingo) [reply Tests/AsyncTests/AsyncTests.swift:30]: @tanner0101 Done! :)

Comment (tanner0101) [reply Tests/AsyncTests/AsyncTests.swift:30]: @adtrevor hmm... did you push the commit? I don't see it.

**Review by calebkleveter — APPROVED** [review]

File: Tests/AsyncTests/AsyncTests.swift (hunk 6 above)
Comment (MrLotU) [inline Tests/AsyncTests/AsyncTests.swift:30]: Still has to be added to the `allTests` :)

**Review by MrLotU — CHANGES_REQUESTED** [review]

**Comment by MrLotU** [comment]
Oh, @adtrevor the tests are kinda borked right now. Could you apply the fixes to the circle file I put in #181 ? That should fix all issues :)

File: Tests/AsyncTests/AsyncTests.swift (hunk 6 above)
Comment (PopFlamingo) [reply Tests/AsyncTests/AsyncTests.swift:30]: @MrLotU Just fixed it

**Review by MrLotU — APPROVED** [review]
LGTM! I'll just have @tanner0101 check it once more and than we can merge :)

**Review by tanner0101 — APPROVED** [review]

**Comment by penny-coin** [comment]
Hey @adtrevor, you just merged a pull request, have a coin! 

You now have 1 coins.

**Comment by PopFlamingo** [comment]
Thank you ! :)

**Pull request merged**